man/hfst-fst2fst.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.3.
.TH HFST-FST2FST "1" "August 2018" "HFST" "User Commands"
.SH NAME
hfst-fst2fst \- =Convert transducers between binary formats
.SH SYNOPSIS
.B hfst-fst2fst
[\fI\,OPTIONS\/\fR...] [\fI\,INFILE\/\fR]
.SH DESCRIPTION
Convert transducers between binary formats
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.SS "Input/Output options:"
.TP
\fB\-i\fR, \fB\-\-input\fR=\fI\,INFILE\/\fR
Read input transducer from INFILE
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,OUTFILE\/\fR
Write output transducer to OUTFILE
.SS "Conversion options:"
.TP
\fB\-f\fR, \fB\-\-format\fR=\fI\,FMT\/\fR
Write result in FMT format
.TP
\fB\-b\fR, \fB\-\-use\-backend\-format\fR
Write result in implementation format, without any HFST wrappers
.TP
\fB\-S\fR, \fB\-\-sfst\fR
Write output in (HFST's) SFST implementation
.TP
\fB\-F\fR, \fB\-\-foma\fR
Write output in (HFST's) foma implementation
.TP
\fB\-x\fR, \fB\-\-xfsm\fR
Write output in native xfsm format
.TP
\fB\-t\fR, \fB\-\-openfst\-tropical\fR
Write output in (HFST's) tropical weight (OpenFST) implementation
.TP
\fB\-l\fR, \fB\-\-openfst\-log\fR
Write output in (HFST's) log weight (OpenFST) implementation
.HP
\fB\-O\fR, \fB\-\-optimized\-lookup\-unweighted\fR Write output in the HFST optimized\-lookup implementation
.TP
\fB\-w\fR, \fB\-\-optimized\-lookup\-weighted\fR
Write output in optimized\-lookup (weighted) implementation
.TP
\fB\-Q\fR  \fB\-\-quick\fR
When converting to optimized\-lookup, don't try hard to compress
.PP
If OUTFILE or INFILE is missing or \-, standard streams will be used.
Format of result depends on format of INFILE
FMT must be name of a format usable by libhfst, i.e. one of the following:
{ foma, openfst\-tropical, openfst\-log, sfst, xfsm
.IP
optimized\-lookup\-weighted, optimized\-lookup\-unweighted }.
.PP
Note that xfsm format is always written in native format without HFST wrappers.
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://github.com/hfst/hfst/issues>
.PP
hfst\-fst2fst home page:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstFst2Fst>
.br
General help using HFST software:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstHome>
.SH COPYRIGHT
Copyright \(co 2017 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
=================================
 NEWS for hfst-tools and library
=================================

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

Noteworthy changes in 3.15.0
----------------------------

* Numerous improvements in pmatch processing:

  * hfst-pmatch2fst now implements eg @include"filename.txt" as a preprocessing stage
  
  * hfst-pmatch2fst now implements Explode() and Implode() in full generality
  
  * hfst-pmatch now allows printing weights (in location mode, with --locate)
  
  * hfst.PmatchContainer.locate offered via python bindings
  
  * bugfixes and improvements hfst.PmatchContainer.tokenize

* Add tool hfst-eliminate-flags.

* Allow both escaped and unescaped zeros in multicharacters in all lexc lexicon entries (fixes issue #274).

* Also reweight state that is both initial and final in hfst-reweight (fixes issue #368).

* Do not let distributed files depend on configuration. Include man files in dist.

* Print warnings and error messages also in earlier stages of twol compilation.

* Add option --execute-and-quit (-E) to hfst-xfst.


Noteworthy changes in 3.14.0
----------------------------

* Numerous improvements to pmatching and tokenization:

  * pmatch now supports the default symbol

  * pmatch now supports reading word embeddings in the binary format

  * improvements to pmatch runtime context handling: a bug affecting expression-initial contexts is fixed, and pmatch now supports Ins() arcs inside RC() and NRC() contexts

  * in pmatch, bugs affecting multiple Ins() arcs, in particular nested ones, are fixed

* Implement variable 'retokenize' in hfst-xfst


Noteworthy changes in 3.13.0
----------------------------

* Numerous improvements to pmatching and tokenization:

  * [].with(X = Y) feature in pmatch This provides support in the pmatch2fst compiler to define "global flags".

  * Add a variable "xerox-composition", default to "on".

  * Consider list symbols (@L..." and "@X...") to be special.

  * Fix runtime handling of contexts and compilation of negative contexts.

  * Make Like() and Unlike() much faster by not sorting the whole vocabulary and only calculating each comparison key once.

  * Keep track of weights along context checking paths and unify weight handling.

  * In blankline-separated mode, keep blanklines in output too.

  * Round weight to zero decimals, non-sci.

  * Make hfst-tokenise usable as a lib; include simple string-to-string function.

  * Use libreadline in hfst-pmatch when available.

  * Cleanup remainder of the pmatch test suite, all the tests now pass.

* Python interface:

  * Support reading several twolc files.

  * Add functions 'compose' and 'cross_product' that take a list of transducers.

  * Allow empty string as input for hfst.fst and hfst.fsa and interpret it as epsilon.

  * Perform fsmbook tests also via python API.

  * Add option --local-hfst to setup.py.

  * Include pre-swig-generated wrappers to pypi source distribution.

* Compilation:

  * Use by default c++11 unordered_map and unordered_set, unless otherwise specified.

  * Add an option --without-c++11 (defaults to 'no') to compile hfst without c++11 support.

  * Require libc++ and osx version >= 10.7 with clang.

* New functions and options:

  * Add function HfstBasicTransducer::remove_final_weight.

  * Add function HfstTransducer::negate() for automata.

  * Add option --restricted-mode (-R) to hfst-xfst.

  * Flag diacritics: support getting a list of operations involved with a particular feature.

  * Allow creating HfstInputStream's from std::istream's.

* Fix issues #341 and #353, make workarounds for issue #358.


Noteworthy changes in 3.12.2
----------------------------

* Changes to configure:

  * Disable lexc and foma wrappers as well as hfst-train-tagger tool unless explicitly requested

  * Enable hfst-calculate and hfst-xfst by default

  * Add experimental options --with-openfst-log=lean and --with-sfst=lean which support a limited nuber of operations for these types (reading, writing, converting between types and deleting)

  * Require at least automake version 1.12 unless compiling from pre-yacc-generated sources

* Implement hfst-twolc as a single program instead of a script. Get rid of hfst-twolc-system and hfst-twolc-loc tools.

* Improve pmatch compilation and error handling mechanisms

* Improve hfst-tokenize tool

* Add transliterate output mode (--transliterate) to hfst-proc

* Changes to python interface:

  * Support twolc and sfst compilation

  * Improve HfstBasicTransducer iteration mechanism

  * Add experimental support for python version 2

  * Support apply up and apply down commands in function start_xfst


Noteworthy changes in 3.12.1
----------------------------

* Fix flag elimination bug (reported in issue #342)

* Do not allow unescaped dots in regular expressions

* Improvements to pmatch and tokenization tools:

  * Search for included files under scriptdir, not working dir

  * Add experimental two-vector model for word sense

  * Handle Apertium-style superblanks in --giella-cg format

* Rename the PyPI package to 'hfst' (currently available at testpypi.python.org)

* Update foma back-end


Noteworthy changes in 3.12.0
----------------------------

* fixes to memory leaks and efficiency

* fixes to numerous warnings

* changes in Python bindings:

  * rename hfst.rules into hfst.sfst_rules

  * get rid of hfst.types and offer implementation types in class ImplementationType

  * add Xerox-type rules in module hfst.xerox_rules

  * improve documentation

  * tentatively add partial support for pypi installation

* improvements to pmatch tools and hfst-proc


Noteworthy changes in 3.11.0
----------------------------

* Add docstrings to Python API

* Changes and improvements to pmatch tools, hfst-tokenize, hfst-optimized-lookup, hfst-lookup and hfst-xfst

* Fix bugs in flag elimination


Noteworthy changes in 3.10.0

tools/src/hfst-fst2fst.cc
//! @file hfst-fst2fst.cc
//!
//! @brief Format conversion command line tool
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#ifdef WINDOWS
#include <io.h>
#endif

#include <iostream>
#include <fstream>

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>

#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "hfst-tool-metadata.h"
#include "HfstTransducer.h"
#include "HfstInputStream.h"
#include "HfstOutputStream.h"

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

using hfst::HfstTransducer;
using hfst::HfstInputStream;
using hfst::HfstOutputStream;
using hfst::ImplementationType;


// tool-specific variables

ImplementationType output_type = hfst::UNSPECIFIED_TYPE;
bool hfst_format = true;
std::string options = "";

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out, "Usage: %s [OPTIONS...] [INFILE]\n"
        "Convert transducers between binary formats\n"
        "\n", program_name);

    print_common_program_options(message_out);
    print_common_unary_program_options(message_out);
    fprintf(message_out, "Conversion options:\n"
    "  -f, --format=FMT                  Write result in FMT format\n"
    "  -b, --use-backend-format          Write result in implementation format, without any HFST wrappers\n"
    "  -S, --sfst                        Write output in (HFST's) SFST implementation\n"
    "  -F, --foma                        Write output in (HFST's) foma implementation\n"
    "  -x, --xfsm                        Write output in native xfsm format\n"
    "  -t, --openfst-tropical            Write output in (HFST's) tropical weight (OpenFST) implementation\n"
    "  -l, --openfst-log                 Write output in (HFST's) log weight (OpenFST) implementation\n"
    "  -O, --optimized-lookup-unweighted Write output in the HFST optimized-lookup implementation\n"
    "  -w, --optimized-lookup-weighted   Write output in optimized-lookup (weighted) implementation\n"
    "  -Q  --quick                       When converting to optimized-lookup, don't try hard to compress\n");
    fprintf(message_out, "\n");
    print_common_unary_program_parameter_instructions(message_out);
        fprintf(message_out, 
            "FMT must be name of a format usable by libhfst, i.e. one of the following:\n"
        "{ foma, openfst-tropical, openfst-log, sfst, xfsm\n"
        "  optimized-lookup-weighted, optimized-lookup-unweighted }.\n"
        "Note that xfsm format is always written in native format without HFST wrappers.\n");
    fprintf(message_out, "\n");
    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
}

int
parse_options(int argc, char** argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
        {
          HFST_GETOPT_COMMON_LONG,
          HFST_GETOPT_UNARY_LONG,
          // add tool-specific options here 
          {"use-backend-format", no_argument, 0, 'b'},
          {"format",       required_argument, 0, 'f'},
          {"sfst",               no_argument, 0, 'S'},
          {"foma",               no_argument, 0, 'F'},
          {"xfsm",               no_argument, 0, 'x'},
          {"openfst-tropical",    no_argument, 0, 't'},
          {"openfst-log",         no_argument, 0, 'l'},
          {"optimized-lookup-unweighted",   no_argument, 0, 'O'},
          {"optimized-lookup-weighted",no_argument, 0, 'w'},
      {"quick",              no_argument, 0, 'Q'},
          {0,0,0,0}
        };
        int option_index = 0;
        // add tool-specific options here 
        char c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT
                             HFST_GETOPT_UNARY_SHORT "SFtlOwQf:bx",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }


        switch (c)
        {
#include "inc/getopt-cases-common.h"
#include "inc/getopt-cases-unary.h"
          // add tool-specific cases here
                case 'f':
          output_type = hfst_parse_format_name(optarg);
          break;
        case 'b':
          hfst_format=false;
          break;
        case 'S':
          output_type = hfst::SFST_TYPE;
          break;
        case 'F':
          output_type = hfst::FOMA_TYPE;
          break;
        case 'x':
          output_type = hfst::XFSM_TYPE;
          break;
        case 't':
          output_type = hfst::TROPICAL_OPENFST_TYPE;
          break;
        case 'l':
          output_type = hfst::LOG_OPENFST_TYPE;
          break;
        case 'O':
          output_type = hfst::HFST_OL_TYPE;
          break;
        case 'w':
          output_type = hfst::HFST_OLW_TYPE;
          break;
    case 'Q':
        options = "quick";
        break;
#include "inc/getopt-cases-error.h"
        }
    }
    
    if (output_type == hfst::UNSPECIFIED_TYPE)
    {
        error(EXIT_FAILURE, 0, 
              "You must specify an output type "
              "(one of -S, -F, -t, -x, -l, -O, or -w)");
    }

#include "inc/check-params-common.h"
#include "inc/check-params-unary.h"
    return EXIT_CONTINUE;
}
int
process_stream(HfstInputStream& instream, HfstOutputStream& outstream)
{
  if (instream.get_type() == hfst::FOMA_TYPE && ! instream.is_hfst_header_included())
    {
      if (!silent)
        {
          fprintf(message_out, "warning: converting native foma transducer: "
                  "inversion may be needed for hfst-lookup to work as expected "
                  "(hfst-flookup works as foma's flookup)\n");
        }
    }
    
    size_t transducer_n = 0;
    while(instream.is_good())
    {
        transducer_n++;
        HfstTransducer orig(instream);

        char* inputname = hfst_get_name(orig, inputfilename);
        if (transducer_n == 1)
        {
          verbose_printf("Converting %s...\n", inputname); 
        }
        else
        {
          verbose_printf("Converting %s..." SIZE_T_SPECIFIER "\n",
                         inputname, transducer_n);
        }
        try {
            orig.convert(output_type, options);
        } HFST_CATCH(HfstException)
        hfst_set_name(orig, orig, "convert");
        hfst_set_formula(orig, orig, "Id");
        outstream << orig;
        free(inputname);
    }
    outstream.flush(); // needed for xfsm transducers whose writing is delayed
    instream.close();
    outstream.close();
    return EXIT_SUCCESS;
}


int main( int argc, char **argv ) {
#ifdef WINDOWS
  _setmode(0, _O_BINARY);
  _setmode(1, _O_BINARY);
#endif
    hfst_set_program_name(argv[0], "0.1", "HfstFst2Fst");
    int retval = parse_options(argc, argv);
    if (retval != EXIT_CONTINUE)
    {
        return retval;
    }
    // close buffers, we use streams
    if (inputfile != stdin)
    {
        fclose(inputfile);
    }
    if (outfile != stdout)
    {
        fclose(outfile);
    }
    verbose_printf("Reading from %s, writing to %s\n", 
        inputfilename, outfilename);
    char* format_description = hfst_strformat(output_type);
    if (hfst_format && (output_type != hfst::XFSM_TYPE)) 
      {
        verbose_printf("Writing %s format transducers with HFST3 headers\n",
                       format_description);
      }
    else
      {
        verbose_printf("Writing %s format transducers without HFST specific"
                       " headers\n", format_description);
      }
    free(format_description);

    if (output_type == hfst::XFSM_TYPE)
      {
        if (strcmp(outfilename, "<stdout>") == 0) {
          error(EXIT_FAILURE, 0, "Writing to standard output not supported for xfsm transducers,\n"
                "use 'hfst-fst2fst [--output|-o] OUTFILE' instead");
          return EXIT_FAILURE;
        }
      }

    // here starts the buffer handling part
    HfstInputStream* instream = NULL;
    try {
      instream = (inputfile != stdin) ?
        new HfstInputStream(inputfilename) : new HfstInputStream();
    } 
    catch(const FileIsInGZFormatException e)
      {
        error(EXIT_FAILURE, 0, "%s seems to be a gzipped native foma file, you must first unzip it",
              inputfilename);
        return EXIT_FAILURE;
      }
    catch(const HfstException e)  {
        error(EXIT_FAILURE, 0, "%s is not a valid transducer file",
              inputfilename);
#if HAVE_XFSM
        if (inputfile == stdin) {
          error(EXIT_FAILURE, 0, "note that you cannot read xfsm transducers from standard input,\n"
                "use hfst-fst2fst [--input|-i] INFILE instead");
        }
#endif
        return EXIT_FAILURE;
    }

    HfstOutputStream* outstream = (outfile != stdout) ?
      new HfstOutputStream(outfilename, output_type, hfst_format) :
      new HfstOutputStream(output_type, hfst_format);

    retval = process_stream(*instream, *outstream);
    delete instream;
    delete outstream;
    free(inputfilename);
    free(outfilename);
    return retval;
}

